### PR TITLE
type(flatMap): should accept fn returning a readonly array

### PR DESCRIPTION
**Pull request by RedHatter**
`flatMap` currently accepts a predicate of `(input: T) => K | Array<K>` however there's no reason the returned array needs to be mutable. This fails when trying to return a readonly array.

This PR updates the predicate type to `(input: T) => K | ReadonlyArray<K>`

---

Make sure that you:

- [x] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [x] New methods added to `src/index.ts`
- [x] New methods added to `mapping.md`

---

<details><summary>We use semantic PR titles to automate the release process!</summary>

https://conventionalcommits.org

PRs should be titled following using the format: `< TYPE >(< scope >)?: description`

### Available Types:

- `feat`: new functions!
- `fix`: changes to function implementations that fix a bug where a function produces the wrong _runtime_ results.
- `type`: type-only changes (params, overloading, return types, etc...)
- `perf`: changes to function implementations that improve a functions _runtime_ performance.
- `refactor`: changes to function implementations that are neither `fix` nor `perf`
- `test`: tests-only changes (transparent to users of the function).
- `docs`: changes to the documentation of a function **or the documentation site**.
- `build`, `ci`, `style`, `chore`, and `revert`: are only relevant for the internals of the library.

For scope put the name of the function you are working on (either new or
existing).

</details>

### CODE REVIEW

File: src/flatMap.test.ts [hunk 1]
```diff
@@ -3,6 +3,14 @@ import { find } from './find';
 import { flatMap } from './flatMap';
 import { pipe } from './pipe';
 
+it('should accept fn returning a readonly array', () => {
+  const result = flatMap(
+    [1, 2] as const,
+    x => [x * 2, x * 3] as ReadonlyArray<number>
```
Comment (eranhirsch) [inline src/flatMap.test.ts:9]: would be enough to just do as const here
```suggestion
    x => [x * 2, x * 3] as const
```

**Review by eranhirsch — CHANGES_REQUESTED** [review]
This looks great:
https://github.com/microsoft/TypeScript/blob/8af8f3c4d1e5cc6654916de9f42e9ba370513157/src/lib/es2019.array.d.ts#L19

Can you also fix the dataLast function signature? you only did the dataFirst one

**Comment by RedHatter** [comment]
Done

**Review by eranhirsch — APPROVED** [review]
Thanks!

**Comment by RedHatter** [comment]
@eranhirsch Can we get this merged? Thanks!

**Comment by eranhirsch** [comment]
Sorry, I don't usually do the merges. It will probably be merged and added to the next version by the end of this week

**Pull request merged**